Request code review of T3 integration

gosha1128 opened the pull request:
@maumueller Per our email conversation, can you validate my various code changes to accomodate T3?  I've tested locally on 4/6 datasets and it looks good so far.

This is currently how I will likely be instructing T3 participants: https://github.com/harsha-simhadri/big-ann-benchmarks/blob/gw/T3/t3/README.md

maumueller commented on benchmark/algorithms/faiss_gpu.py:
```diff
@@ -0,0 +1,108 @@
+        return f"data/ivf_{name}_{self._n_list}_{self._metric}"
+
+    def fit(self, dataset):
+        X = DATASETS[dataset].get_dataset() 
```
The gpu variant is another baseline for in-memory datasets, right?

gosha1128 replied:
This was my first attempt at a faiss gpu algorithm.  Obviously, we moved to the index factory - so I will remove this one.

maumueller commented on benchmark/algorithms/faiss_t3.py:
```diff
@@ -0,0 +1,404 @@
+            yield res
+        res = res_next
+
+def build_index(buildthreads, by_residual, maxtrain, clustering_niter, 
+     indexkey, indexfile, add_bs, add_splits, ds, train_on_gpu=True, quantizer_on_gpu_add=True):
```
I wonder if T1 and T3 should re-use the same build_index code? It looks almost c&p?

gosha1128 replied:
I think you are right.  It looks like the only major difference between T1 and T3 is the use of the gpu-quantizer.  Perhaps a new base class could accomplish this.

maumueller commented on benchmark/algorithms/faiss_t3.py:
```diff
@@ -0,0 +1,404 @@
+        ds = DATASETS[dataset]()
+        d = ds.d
+
+        buildthreads = index_params.get("buildthreads", -1)
+        by_residual = index_params.get("by_residual", -1) 
+        maxtrain = index_params.get("maxtrain", 0)
+        clustering_niter = index_params.get("clustering_niter", -1)
+        indexkey = index_params.get("indexkey", "IVF1048576,SQ8")
+        add_bs = index_params.get("add_bs", 100000)
+        add_splits = index_params.get("add_splits", 1)
+        indexfile = self.index_name(dataset)
```
Very nice, let me use this approach for T1 as well.

maumueller commented on benchmark/datasets.py:
```diff
@@ -85,6 +85,21 @@ def xbin_mmap(fname, dtype, maxn=-1):
         n = min(n, maxn)
     return np.memmap(fname, dtype=dtype, mode="r", offset=8, shape=(n, d))
 
+#GW - I may eventually remove these
+def xbin_mmap_floatarray(fname, shape):
+    x = np.memmap(fname, dtype=numpy.float32, mode="r", offset=0, shape=shape)
+    return x
+
+def xbin_mmap_with_cast(fname, r_dtype, w_dtype, maxn=-1):
+    """ mmap the competition file format for a given type of items """
+    n, d = map(int, np.fromfile(fname, dtype="uint32", count=2))
+    assert os.stat(fname).st_size == 8 + n * d * np.dtype(r_dtype).itemsize
+    n = max(n, maxn)
+    x = np.memmap(fname, dtype=r_dtype, mode="r", offset=8, shape=(n, d))
+    x = x.astype( w_dtype )
+    return x
+#GW
```
These seem to be used in the dataset creation. I'm not sure why T1/T2/T3 should use different abstractions on the datasets?

gosha1128 replied:
These were useful to me at some point, but not needed anymore.  Removing them.

maumueller commented on benchmark/datasets.py:
```diff
@@ -369,6 +384,75 @@ def __init__(self, nb_M=1000):
+    def get_groundtruth(self,k=None):
+        if self.gt_fn:
+            return super().get_groundtruth(k)
+        else: # assume its a crop
+            gt_crop = os.path.join( self.basedir, "gt.ibin.crop_nb_%d" % self.nb )
+            if os.path.exists(gt_crop):
+                gt = xbin_mmap(os.path.join( gt_crop), dtype=np.int32)
+                if k is not None:
+                    assert k <= 100
+                    gt = gt[:, :k]
+                return gt
+            else:
+                raise Exception("No ground truth file.")
+
+    def get_queries(self):
+        ds_fn = super().get_dataset_fn()
+        ds_fn_fp = ds_fn + ".float32"
+        if os.path.exists( ds_fn_fp ):
+            filename = os.path.join(self.basedir, self.qs_fn)
+            return xbin_mmap_with_cast( filename, np.uint8, np.float32 )
+        else:
+            return super().get_queries()
+
+    def get_dataset_iterator(self, bs=512, split=(1,0)):
+        ds_fn = super().get_dataset_fn()
+        ds_fn_fp = ds_fn + ".float32"
+        if os.path.exists( ds_fn_fp ):
+            nsplit, rank = split
+            i0, i1 = self.nb * rank // nsplit, self.nb * (rank + 1) // nsplit
+            filename = self.get_dataset_fn()
+            x = xbin_mmap_floatarray(filename, (self.nb, 128))
+            assert x.shape == (self.nb, self.d)
+            for j0 in range(i0, i1, bs):
+                j1 = min(j0 + bs, i1)
+                yield sanitize(x[j0:j1])
+        else:
+            return super().get_dataset_iterator(bs, split)
+
+    def distance(self):
+        return "euclidean"
```
Why is this necessary? Is this for debugging?

gosha1128 replied:
Yeah.  It was for debugging.  Removing.

maumueller commented on benchmark/datasets.py:
```diff
@@ -608,14 +696,17 @@ def __str__(self):
     'msspacev-100M': lambda : MSSPACEV1B(100),
     'msspacev-1M': lambda : MSSPACEV1B(1),
 
+    'random-xs': lambda: RandomDS(10000, 1000, 20),
+    'random-s': lambda : RandomDS(100000, 1000, 50),
+
     'random-xs': lambda : RandomDS(10000, 1000, 20),
     'random-s': lambda : RandomDS(100000, 1000, 50),
 
```
This seems to be a merge error?

gosha1128 replied:
Yep, thanks for catching.

maumueller commented on benchmark/datasets.py.gw:
```diff
@@ -0,0 +1,713 @@
+import math
```
Can we leave this file out of the PR?

gosha1128 replied:
Oops, removing.

maumueller commented on benchmark/plotting/utils.py:
```diff
@@ -73,6 +73,47 @@ def compute_metrics_all_runs(dataset, res, recompute=False):
+            run_result[name] = v
+        yield run_result
+
+def compute_metrics_all_runs(dataset, res, recompute=False):
```
This looks like a merge problem. Can you check?

gosha1128 replied:
Yep.

maumueller commented on benchmark/runner.py:
```diff
@@ -99,13 +100,18 @@ def run(definition, dataset, count, run_count, rebuild):
 
 
-def run_from_cmdline():
+def run_from_cmdline(args=None):
```
I'm not sure if this is very clever or very dangerous in the long run :-)

gosha1128 replied:
Probably more the latter :)

maumueller commented on benchmark/t3/helper.py:
```diff
@@ -0,0 +1,122 @@
+    else:
+        raise Exception("Docker invoke not supported for this algorithm.")
+
+def create_container_with_gpu_support(client, image, command, volumes, **kwargs):
+
+
+    from docker.models.images import Image
+    from docker.models.containers import _create_container_args
+
+    if isinstance(image, Image):
+        image = image.id
+
+    os.environ['NVIDIA_VISIBLE_DEVICES']='all'
+
+    kwargs['image'] = image
+    kwargs['command'] = command
+    kwargs['version'] = client.containers.client.api._version
+    kwargs['volumes'] = volumes
+    create_kwargs = _create_container_args(kwargs)
+
+    device_request = {
+        'Driver': 'nvidia',
+        'Capabilities': [['gpu'], ['nvidia']],
+        'Count': -1,  # enable all gpus
+    }
+
+    if device_request is not None:
+        create_kwargs['host_config']['DeviceRequests'] = [device_request]
+
+    resp = client.api.create_container(**create_kwargs)
+    return client.containers.get(resp['Id'])
```
Awesome! I didn't know you could do that.

gosha1128 replied:
Yes, in order to use the nvidia docker cuda run-time - you need to do some weird tricks.

maumueller commented on benchmark/t3/helper.py:
```diff
@@ -0,0 +1,122 @@
+    def distance(self):
+        return "angular"
+
+def t3_augment_datasets():
+    DATASETS.update(
+    {
+        #'sift-1M-angular': BigANNDatasetAngular(1),
+        'random-s-256d':  RandomDS(100000, 1000, 256),
+        'random-m-256d':  RandomDS(1000000, 1000, 256),
+        'random-l-256d':  RandomDS(10000000, 1000, 256),
+        'random-ml-128d': RandomDS(50000000, 1000, 128),
+        'random-xl-64d': RandomDS(100000000, 1000, 64)
+    })
+
+
```
If those were useful, I suggest adding them to `datasets.py`.

gosha1128 replied:
Not really useful any longer, removing.

maumueller commented on create_dataset.py.gw:
```diff
@@ -0,0 +1,12 @@
+import argparse
```
Can we leave this file out?

gosha1128 replied:
Oops, removing.

maumueller commented on install.py:
```diff
@@ -5,17 +5,19 @@
 
 
-def build(library, args):
+def build(library, args, dockerfile):
```
Isn't dockerfile identical to args.algorithm? (Maybe the latter has a confusing name?) Or do you really want to point to dockerfile that are not present in `install/`?

gosha1128 replied:
I think my rationale was to keep T3 docker builds out of the default when you run "python install.py."  Is there a better way to accomplish this?

maumueller replied:
I see, let's keep it like you proposed.

maumueller commented on t3/README.md:
```diff
@@ -0,0 +1,197 @@
+# T3 Track
```
Very nice! Many parts for the "for participants" could be shared among the three tracks. Let's do this after merging this PR.

maumueller commented on t3/faiss_t3/algos.yaml:
```diff
@@ -0,0 +1,244 @@
+        base:
+          args: |
+              [{"indexkey": "IVF1048576,SQ8",
+                "buildthreads": -1,
+                "by_residual": -1,
+                "add_bs": 100000,
+                "maxtrain": 0,
+                "clustering_niter": -1,
+                "add_splits": 1
```
These seem to be the default parameters, so you could tidy up a bit and leave it only in one place to show the possible configurations.

gosha1128 replied:
Yep. I'll do exactly what you do in the toplevel algos.yaml.

maumueller commented on t3/gpu_baseline_faiss.py:
```diff
@@ -0,0 +1,592 @@
+    if index_ivf is not None:
+        print("imbalance_factor=", index_ivf.invlists.imbalance_factor())
+
+    if args.no_precomputed_tables:
+        if isinstance(index_ivf, faiss.IndexIVFPQ):
+            print("disabling precomputed table")
+            index_ivf.use_precomputed_table = -1
+            index_ivf.precomputed_table.clear()
+
+    if args.indexfile:
+        print("index size on disk: ", os.stat(args.indexfile).st_size)
+
+    print("current RSS:", faiss.get_mem_usage_kb() * 1024)
+
+    precomputed_table_size = 0
+    if hasattr(index_ivf, 'precomputed_table'):
+        precomputed_table_size = index_ivf.precomputed_table.size() * 4
+
+    print("precomputed tables size:", precomputed_table_size)
+
+    if args.search:
+
+        if args.searchthreads == -1:
+            print("Search threads:", faiss.omp_get_max_threads())
+        else:
+            print("Setting nb of threads to", args.searchthreads)
+            faiss.omp_set_num_threads(args.searchthreads)
+
+        if args.parallel_mode != -1:
+            print("setting IVF parallel mode to", args.parallel_mode)
+            index_ivf.parallel_mode
+            index_ivf.parallel_mode = args.parallel_mode
+
+        if args.searchparams == ["autotune"]:
+            run_experiments_autotune(ds, index, args)
+        else:
+            run_experiments_searchparams(ds, index, args ) #, 96237 )
+
+if __name__ == "__main__":
+    main()
```
Full support for leaving this baseline in, but we should add a comment that this is stand-alone and doesn't use the framework (very much).

gosha1128 replied:
Actually, i'm fine removing it.  It would just cause unnecessary confusion.

maumueller reviewed (commented):
Great job, @gosha1128. I only have a few very minor remarks, but this looks good to me on a code level.

gosha1128 commented:
Great review- thanks.

I’ll make the necessary modifications and ( if you don’t mind ) I’d like to
resubmit the PR to you.

George

On Tue, Jul 27, 2021 at 12:57 AM Martin Aumüller ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> Great job, @gosha1128 <https://github.com/gosha1128>. I only have a few
> very minor remarks, but this looks good to me on a code level.
> ------------------------------
>
> In benchmark/algorithms/faiss_gpu.py
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#discussion_r677190950>
> :
>
> > +        self._metric = metric
> +        self.index = None
> +        self.gindex = None
> +        self.quantizer = None
> +        self.res = None
> +        self.resource = None
> +
> +    def done(self):
> +        if self.gindex:
> +            self.gindex.reset() # release GPU and memory?
> +
> +    def index_name(self, name):
> +        return f"data/ivf_{name}_{self._n_list}_{self._metric}"
> +
> +    def fit(self, dataset):
> +        X = DATASETS[dataset].get_dataset()
>
> The gpu variant is another baseline for in-memory datasets, right?
> ------------------------------
>
> In benchmark/algorithms/faiss_t3.py
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#discussion_r677191932>
> :
>
> > +def build_index(buildthreads, by_residual, maxtrain, clustering_niter,
> +     indexkey, indexfile, add_bs, add_splits, ds, train_on_gpu=True, quantizer_on_gpu_add=True):
>
> I wonder if T1 and T3 should re-use the same build_index code? It looks
> almost c&p?
> ------------------------------
>
> In benchmark/algorithms/faiss_t3.py
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#discussion_r677192565>
> :
>
> > +        buildthreads = index_params.get("buildthreads", -1)
> +        by_residual = index_params.get("by_residual", -1)
> +        maxtrain = index_params.get("maxtrain", 0)
> +        clustering_niter = index_params.get("clustering_niter", -1)
> +        indexkey = index_params.get("indexkey", "IVF1048576,SQ8")
> +        add_bs = index_params.get("add_bs", 100000)
> +        add_splits = index_params.get("add_splits", 1)
> +        indexfile = self.index_name(dataset)
>
> Very nice, let me use this approach for T1 as well.
> ------------------------------
>
> In benchmark/datasets.py
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#discussion_r677195162>
> :
>
> > +#GW - I may eventually remove these
> +def xbin_mmap_floatarray(fname, shape):
> +    x = np.memmap(fname, dtype=numpy.float32, mode="r", offset=0, shape=shape)
> +    return x
> +
> +def xbin_mmap_with_cast(fname, r_dtype, w_dtype, maxn=-1):
> +    """ mmap the competition file format for a given type of items """
> +    n, d = map(int, np.fromfile(fname, dtype="uint32", count=2))
> +    assert os.stat(fname).st_size == 8 + n * d * np.dtype(r_dtype).itemsize
> +    n = max(n, maxn)
> +    x = np.memmap(fname, dtype=r_dtype, mode="r", offset=8, shape=(n, d))
> +    x = x.astype( w_dtype )
> +    return x
> +#GW
>
> These seem to be used in the dataset creation. I'm not sure why T1/T2/T3
> should use different abstractions on the datasets?
> ------------------------------
>
> In benchmark/datasets.py
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#discussion_r677196468>
> :
>
> > +class BigANNDatasetT3(DatasetCompetitionFormat):
> +    def __init__(self, nb_M=1000):
> +        self.nb_M = nb_M
> +        self.nb = 10**6 * nb_M
> +        self.d = 128
> +        self.nq = 10000
> +        self.dtype = "uint8"
> +        self.ds_fn = "base.1B.u8bin"
> +        self.qs_fn = "query.public.10K.u8bin"
> +        self.gt_fn = "GT.public.1B.ibin" if self.nb == 10**9 else None
> +        self.base_url = "https://dl.fbaipublicfiles.com/billion-scale-ann-benchmarks/bigann/"
> +        self.basedir = os.path.join(BASEDIR, "bigann")
> +
> +    def get_dataset_fn(self):
> +        ds_fn = super().get_dataset_fn()
> +        ds_fn_fp = ds_fn + ".float32"
> +        if os.path.exists( ds_fn_fp ):
> +            return ds_fn_fp
> +        else:
> +            return ds_fn
> +
> +    def get_dataset(self):
> +        ds_fn = super().get_dataset_fn()
> +        ds_fn_fp = ds_fn + ".float32"
> +        if os.path.exists( ds_fn_fp ):
> +            return xbin_mmap_floatarray( ds_fn_fp, ( self.nb, self.d ) )
> +        else:
> +            return super().get_dataset()
> +
> +    def get_groundtruth(self,k=None):
> +        if self.gt_fn:
> +            return super().get_groundtruth(k)
> +        else: # assume its a crop
> +            gt_crop = os.path.join( self.basedir, "gt.ibin.crop_nb_%d" % self.nb )
> +            if os.path.exists(gt_crop):
> +                gt = xbin_mmap(os.path.join( gt_crop), dtype=np.int32)
> +                if k is not None:
> +                    assert k <= 100
> +                    gt = gt[:, :k]
> +                return gt
> +            else:
> +                raise Exception("No ground truth file.")
> +
> +    def get_queries(self):
> +        ds_fn = super().get_dataset_fn()
> +        ds_fn_fp = ds_fn + ".float32"
> +        if os.path.exists( ds_fn_fp ):
> +            filename = os.path.join(self.basedir, self.qs_fn)
> +            return xbin_mmap_with_cast( filename, np.uint8, np.float32 )
> +        else:
> +            return super().get_queries()
> +
> +    def get_dataset_iterator(self, bs=512, split=(1,0)):
> +        ds_fn = super().get_dataset_fn()
> +        ds_fn_fp = ds_fn + ".float32"
> +        if os.path.exists( ds_fn_fp ):
> +            nsplit, rank = split
> +            i0, i1 = self.nb * rank // nsplit, self.nb * (rank + 1) // nsplit
> +            filename = self.get_dataset_fn()
> +            x = xbin_mmap_floatarray(filename, (self.nb, 128))
> +            assert x.shape == (self.nb, self.d)
> +            for j0 in range(i0, i1, bs):
> +                j1 = min(j0 + bs, i1)
> +                yield sanitize(x[j0:j1])
> +        else:
> +            return super().get_dataset_iterator(bs, split)
> +
> +    def distance(self):
> +        return "euclidean"
>
> Why is this necessary? Is this for debugging?
> ------------------------------
>
> In benchmark/datasets.py
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#discussion_r677196830>
> :
>
> > +    'random-xs': lambda: RandomDS(10000, 1000, 20),
> +    'random-s': lambda : RandomDS(100000, 1000, 50),
> +
>      'random-xs': lambda : RandomDS(10000, 1000, 20),
>      'random-s': lambda : RandomDS(100000, 1000, 50),
>
>
> This seems to be a merge error?
> ------------------------------
>
> In benchmark/datasets.py.gw
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#discussion_r677197309>
> :
>
> > @@ -0,0 +1,713 @@
> +import math
>
> Can we leave this file out of the PR?
> ------------------------------
>
> In benchmark/plotting/utils.py
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#discussion_r677198702>
> :
>
> > +        except:
> +            pass
> +
> +        run_result = {
> +            'algorithm': algo,
> +            'parameters': algo_name,
> +            'dataset': dataset,
> +            'count': properties['count'],
> +        }
> +        for name, metric in metrics.items():
> +            if not power_capture.run_has_power_stats(properties): continue
> +            v = metric["function"](true_nn, run_nn, metrics_cache, properties)
> +            run_result[name] = v
> +        yield run_result
> +
> +def compute_metrics_all_runs(dataset, res, recompute=False):
>
> This looks like a merge problem. Can you check?
> ------------------------------
>
> In benchmark/runner.py
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#discussion_r677200214>
> :
>
> >      finally:
>          algo.done()
>
>
> -def run_from_cmdline():
> +def run_from_cmdline(args=None):
>
> I'm not sure if this is very clever or very dangerous in the long run :-)
> ------------------------------
>
> In benchmark/t3/helper.py
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#discussion_r677202691>
> :
>
> > +def create_container_with_gpu_support(client, image, command, volumes, **kwargs):
> +
> +
> +    from docker.models.images import Image
> +    from docker.models.containers import _create_container_args
> +
> +    if isinstance(image, Image):
> +        image = image.id
> +
> +    os.environ['NVIDIA_VISIBLE_DEVICES']='all'
> +
> +    kwargs['image'] = image
> +    kwargs['command'] = command
> +    kwargs['version'] = client.containers.client.api._version
> +    kwargs['volumes'] = volumes
> +    create_kwargs = _create_container_args(kwargs)
> +
> +    device_request = {
> +        'Driver': 'nvidia',
> +        'Capabilities': [['gpu'], ['nvidia']],
> +        'Count': -1,  # enable all gpus
> +    }
> +
> +    if device_request is not None:
> +        create_kwargs['host_config']['DeviceRequests'] = [device_request]
> +
> +    resp = client.api.create_container(**create_kwargs)
> +    return client.containers.get(resp['Id'])
>
> Awesome! I didn't know you could do that.
> ------------------------------
>
> In benchmark/t3/helper.py
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#discussion_r677204295>
> :
>
> > +def t3_augment_datasets():
> +    DATASETS.update(
> +    {
> +        #'sift-1M-angular': BigANNDatasetAngular(1),
> +        'random-s-256d':  RandomDS(100000, 1000, 256),
> +        'random-m-256d':  RandomDS(1000000, 1000, 256),
> +        'random-l-256d':  RandomDS(10000000, 1000, 256),
> +        'random-ml-128d': RandomDS(50000000, 1000, 128),
> +        'random-xl-64d': RandomDS(100000000, 1000, 64)
> +    })
> +
> +
>
> If those were useful, I suggest adding them to datasets.py.
> ------------------------------
>
> In create_dataset.py.gw
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#discussion_r677204381>
> :
>
> > @@ -0,0 +1,12 @@
> +import argparse
>
> Can we leave this file out?
> ------------------------------
>
> In install.py
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#discussion_r677205659>
> :
>
> > @@ -5,17 +5,19 @@
>  from multiprocessing import Pool
>
>
> -def build(library, args):
> +def build(library, args, dockerfile):
>
> Isn't dockerfile identical to args.algorithm? (Maybe the latter has a
> confusing name?) Or do you really want to point to dockerfile that are not
> present in install/?
> ------------------------------
>
> In t3/README.md
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#discussion_r677207377>
> :
>
> > @@ -0,0 +1,197 @@
> +# T3 Track
>
> Very nice! Many parts for the "for participants" could be shared among the
> three tracks. Let's do this after merging this PR.
> ------------------------------
>
> In t3/faiss_t3/algos.yaml
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#discussion_r677208548>
> :
>
> > +                "buildthreads": -1,
> +                "by_residual": -1,
> +                "add_bs": 100000,
> +                "maxtrain": 0,
> +                "clustering_niter": -1,
> +                "add_splits": 1
>
> These seem to be the default parameters, so you could tidy up a bit and
> leave it only in one place to show the possible configurations.
> ------------------------------
>
> In t3/gpu_baseline_faiss.py
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#discussion_r677209707>
> :
>
> > +        resource.setrlimit(
> +            resource.RLIMIT_DATA, (args.maxRAM * 1024 ** 3, resource.RLIM_INFINITY)
> +        )
> +
> +    os.system('echo -n "nb processors "; '
> +          'cat /proc/cpuinfo | grep ^processor | wc -l; '
> +          'cat /proc/cpuinfo | grep ^"model name" | tail -1')
> +
> +
> +    #GW ds = DATASETS[args.dataset]()
> +    ds = DATASETS[args.dataset]
> +
> +    if args.search:
> +        # Do a sanity check before loading a large index file
> +        q = ds.get_queries()
> +        #range search- r, i, g = ds.get_groundtruth()
> +        i, g = ds.get_groundtruth()
> +
> +    if args.build:
> +        # Do a sanity check before training large dataset
> +        d = ds.get_dataset()
> +        for xblock in ds.get_dataset_iterator(bs=10):
> +            break
> +
> +    if args.search:
> +
> +        # Connect to ipmicap server and do a test
> +        power_capture("192.168.99.123:1234:-1")
> +        power_capture.ping()
> +        capid = power_capture.start()
> +        time.sleep(2)
> +        power_cons = power_capture.stop(capid, all_stats=True)
> +        if power_cons==-1: raise Exception("Power capture failed.")
> +        print("Power sensor test passed=", power_cons)
> +
> +        # Initialize the data files
> +        init_data(args)
> +
> +    nq, d = ds.nq, ds.d
> +    nb, d = ds.nq, ds.d
> +
> +    if args.prepare:
> +        print("downloading dataset...")
> +        ds.prepare()
> +        print("dataset ready")
> +
> +    if not (args.build or args.search):
> +        return
> +
> +    if args.build:
> +        print("build index, key=", args.indexkey)
> +        index = build_index(args, ds)
> +    else:
> +        print("reading", args.indexfile)
> +        index = faiss.read_index(args.indexfile)
> +
> +    index_ivf, vec_transform = unwind_index_ivf(index)
> +    if vec_transform is None:
> +        vec_transform = lambda x: x
> +
> +    if index_ivf is not None:
> +        print("imbalance_factor=", index_ivf.invlists.imbalance_factor())
> +
> +    if args.no_precomputed_tables:
> +        if isinstance(index_ivf, faiss.IndexIVFPQ):
> +            print("disabling precomputed table")
> +            index_ivf.use_precomputed_table = -1
> +            index_ivf.precomputed_table.clear()
> +
> +    if args.indexfile:
> +        print("index size on disk: ", os.stat(args.indexfile).st_size)
> +
> +    print("current RSS:", faiss.get_mem_usage_kb() * 1024)
> +
> +    precomputed_table_size = 0
> +    if hasattr(index_ivf, 'precomputed_table'):
> +        precomputed_table_size = index_ivf.precomputed_table.size() * 4
> +
> +    print("precomputed tables size:", precomputed_table_size)
> +
> +    if args.search:
> +
> +        if args.searchthreads == -1:
> +            print("Search threads:", faiss.omp_get_max_threads())
> +        else:
> +            print("Setting nb of threads to", args.searchthreads)
> +            faiss.omp_set_num_threads(args.searchthreads)
> +
> +        if args.parallel_mode != -1:
> +            print("setting IVF parallel mode to", args.parallel_mode)
> +            index_ivf.parallel_mode
> +            index_ivf.parallel_mode = args.parallel_mode
> +
> +        if args.searchparams == ["autotune"]:
> +            run_experiments_autotune(ds, index, args)
> +        else:
> +            run_experiments_searchparams(ds, index, args ) #, 96237 )
> +
> +if __name__ == "__main__":
> +    main()
>
> Full support for leaving this baseline in, but we should add a comment
> that this is stand-alone and doesn't use the framework (very much).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#pullrequestreview-715576447>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADL6CIOSX5UUQ77ATZPMETTZZRFRANCNFSM5BBCWJOQ>
> .
>

maumueller commented on t3/faiss_t3/algos.yaml:
```diff
@@ -142,16 +119,9 @@ msturing-10M:
+              [{"indexkey": "IVF1048576,SQ8"}]
           query-args: |
-              ["nprobe=1",
+              ["nprobe=1"]
```
Is this intentional?

gosha1128 replied:
Yes, that is fixed now.

maumueller commented:
What do you think about presenting all the baseline results in one plot in the `results/` folder? I think @harsha-simhadri  has the results for T1, so he would have to share them first.

maumueller commented:
Running this on a local server, I get 

```
maau@preciousss:/mnt/large_storage/maau/big-ann-benchmarks$ python3 run.py --dataset deep-10M --algorithm faiss-t3
file data/deep1b/query.public.10K.fbin already exists
file data/deep1b/deep-10M already exists
2021-07-28 12:54:35,235 - annb - INFO - running only faiss-t3
2021-07-28 12:54:35,492 - annb - INFO - Order: [Definition(algorithm='faiss-t3', constructor='FaissT3', module='benchmark.algorithms.faiss_t3', docker_tag='billion-scale-benchmark-faiss_t3', arguments=['euclidean', {'indexkey': 'IVF65536,SQ8', 'buildthreads': -1, 'by_residual': -1, 'add_bs': 100000, 'maxtrain': 0, 'clustering_niter': -1, 'add_splits': 1}], query_argument_groups=[['nprobe=1'], ['nprobe=2'], ['nprobe=4'], ['nprobe=8'], ['nprobe=16'], ['nprobe=32'], ['nprobe=64'], ['nprobe=128'], ['nprobe=256']], disabled=False)]
2021-07-28 12:54:36,152 - annb.319d272f67 - INFO - Created container 319d272f67: CPU limit 0-55, mem limit 535741299456, timeout 43200, command ['--dataset', 'deep-10M', '--algorithm', 'faiss-t3', '--module', 'benchmark.algorithms.faiss_t3', '--constructor', 'FaissT3', '--runs', '5', '--count', '10', '["euclidean", {"indexkey": "IVF65536,SQ8", "buildthreads": -1, "by_residual": -1, "add_bs": 100000, "maxtrain": 0, "clustering_niter": -1, "add_splits": 1}]', '["nprobe=1"]', '["nprobe=2"]', '["nprobe=4"]', '["nprobe=8"]', '["nprobe=16"]', '["nprobe=32"]', '["nprobe=64"]', '["nprobe=128"]', '["nprobe=256"]']
2021-07-28 12:54:40,490 - annb.319d272f67 - INFO - Faiss assertion 'err__ == cudaSuccess' failed in int faiss::gpu::getNumDevices() at /root/miniconda3/conda-bld/faiss-pkg_1623030479928/work/faiss/gpu/utils/DeviceUtils.cu:36; details: CUDA error 35 CUDA driver version is insufficient for CUDA runtime version
2021-07-28 12:54:41,094 - annb.319d272f67 - ERROR - Faiss assertion 'err__ == cudaSuccess' failed in int faiss::gpu::getNumDevices() at /root/miniconda3/conda-bld/faiss-pkg_1623030479928/work/faiss/gpu/utils/DeviceUtils.cu:36; details: CUDA error 35 CUDA driver version is insufficient for CUDA runtime version

2021-07-28 12:54:41,094 - annb.319d272f67 - ERROR - Child process for container 319d272f67 raised exception 139
```

with host having `NVIDIA-SMI 470.42.01    Driver Version: 470.42.01    CUDA Version: 11.4`. Would I have to update the dockerfile to point at more recent version? If it runs in your setup, I'm fine with merging.

gosha1128 commented:
Yeah.  I should put the full installation instructions for the FAISS T3
baseline into the README ( it's just a light mention right now. )

My setup is 1) standard CUDA 11.0 drivers, and 2) the CUDA 11.0
docker/container run time:
https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html

Note that if you want to try without docker (ie, you just have the CUDA
drivers), I added a "--nodocker" flag to run.py.  This was useful during
debugging.

George

On Wed, Jul 28, 2021 at 4:04 AM Martin Aumüller ***@***.***>
wrote:

> Running this on a local server, I get
>
> ***@***.***:/mnt/large_storage/maau/big-ann-benchmarks$ python3 run.py --dataset deep-10M --algorithm faiss-t3
> file data/deep1b/query.public.10K.fbin already exists
> file data/deep1b/deep-10M already exists
> 2021-07-28 12:54:35,235 - annb - INFO - running only faiss-t3
> 2021-07-28 12:54:35,492 - annb - INFO - Order: [Definition(algorithm='faiss-t3', constructor='FaissT3', module='benchmark.algorithms.faiss_t3', docker_tag='billion-scale-benchmark-faiss_t3', arguments=['euclidean', {'indexkey': 'IVF65536,SQ8', 'buildthreads': -1, 'by_residual': -1, 'add_bs': 100000, 'maxtrain': 0, 'clustering_niter': -1, 'add_splits': 1}], query_argument_groups=[['nprobe=1'], ['nprobe=2'], ['nprobe=4'], ['nprobe=8'], ['nprobe=16'], ['nprobe=32'], ['nprobe=64'], ['nprobe=128'], ['nprobe=256']], disabled=False)]
> 2021-07-28 12:54:36,152 - annb.319d272f67 - INFO - Created container 319d272f67: CPU limit 0-55, mem limit 535741299456, timeout 43200, command ['--dataset', 'deep-10M', '--algorithm', 'faiss-t3', '--module', 'benchmark.algorithms.faiss_t3', '--constructor', 'FaissT3', '--runs', '5', '--count', '10', '["euclidean", {"indexkey": "IVF65536,SQ8", "buildthreads": -1, "by_residual": -1, "add_bs": 100000, "maxtrain": 0, "clustering_niter": -1, "add_splits": 1}]', '["nprobe=1"]', '["nprobe=2"]', '["nprobe=4"]', '["nprobe=8"]', '["nprobe=16"]', '["nprobe=32"]', '["nprobe=64"]', '["nprobe=128"]', '["nprobe=256"]']
> 2021-07-28 12:54:40,490 - annb.319d272f67 - INFO - Faiss assertion 'err__ == cudaSuccess' failed in int faiss::gpu::getNumDevices() at /root/miniconda3/conda-bld/faiss-pkg_1623030479928/work/faiss/gpu/utils/DeviceUtils.cu:36; details: CUDA error 35 CUDA driver version is insufficient for CUDA runtime version
> 2021-07-28 12:54:41,094 - annb.319d272f67 - ERROR - Faiss assertion 'err__ == cudaSuccess' failed in int faiss::gpu::getNumDevices() at /root/miniconda3/conda-bld/faiss-pkg_1623030479928/work/faiss/gpu/utils/DeviceUtils.cu:36; details: CUDA error 35 CUDA driver version is insufficient for CUDA runtime version
>
> 2021-07-28 12:54:41,094 - annb.319d272f67 - ERROR - Child process for container 319d272f67 raised exception 139
>
> with host having NVIDIA-SMI 470.42.01 Driver Version: 470.42.01 CUDA
> Version: 11.4. Would I have to update the dockerfile to point at more
> recent version? If it runs in your setup, I'm fine with merging.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#issuecomment-888220165>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADL6CJADYMMTXRDE7TAVADTZ7P3VANCNFSM5BBCWJOQ>
> .
>

maumueller commented:
> Yeah. I should put the full installation instructions for the FAISS T3 baseline into the README ( it's just a light mention right now. ) My setup is 1) standard CUDA 11.0 drivers, and 2) the CUDA 11.0 docker/container run time: https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html Note that if you want to try without docker (ie, you just have the CUDA drivers), I added a "--nodocker" flag to run.py. This was useful during debugging. George
> […](#)
> On Wed, Jul 28, 2021 at 4:04 AM Martin Aumüller ***@***.***> wrote: Running this on a local server, I get ***@***.***:/mnt/large_storage/maau/big-ann-benchmarks$ python3 run.py --dataset deep-10M --algorithm faiss-t3 file data/deep1b/query.public.10K.fbin already exists file data/deep1b/deep-10M already exists 2021-07-28 12:54:35,235 - annb - INFO - running only faiss-t3 2021-07-28 12:54:35,492 - annb - INFO - Order: [Definition(algorithm='faiss-t3', constructor='FaissT3', module='benchmark.algorithms.faiss_t3', docker_tag='billion-scale-benchmark-faiss_t3', arguments=['euclidean', {'indexkey': 'IVF65536,SQ8', 'buildthreads': -1, 'by_residual': -1, 'add_bs': 100000, 'maxtrain': 0, 'clustering_niter': -1, 'add_splits': 1}], query_argument_groups=[['nprobe=1'], ['nprobe=2'], ['nprobe=4'], ['nprobe=8'], ['nprobe=16'], ['nprobe=32'], ['nprobe=64'], ['nprobe=128'], ['nprobe=256']], disabled=False)] 2021-07-28 12:54:36,152 - annb.319d272f67 - INFO - Created container 319d272f67: CPU limit 0-55, mem limit 535741299456, timeout 43200, command ['--dataset', 'deep-10M', '--algorithm', 'faiss-t3', '--module', 'benchmark.algorithms.faiss_t3', '--constructor', 'FaissT3', '--runs', '5', '--count', '10', '["euclidean", {"indexkey": "IVF65536,SQ8", "buildthreads": -1, "by_residual": -1, "add_bs": 100000, "maxtrain": 0, "clustering_niter": -1, "add_splits": 1}]', '["nprobe=1"]', '["nprobe=2"]', '["nprobe=4"]', '["nprobe=8"]', '["nprobe=16"]', '["nprobe=32"]', '["nprobe=64"]', '["nprobe=128"]', '["nprobe=256"]'] 2021-07-28 12:54:40,490 - annb.319d272f67 - INFO - Faiss assertion 'err__ == cudaSuccess' failed in int faiss::gpu::getNumDevices() at /root/miniconda3/conda-bld/faiss-pkg_1623030479928/work/faiss/gpu/utils/DeviceUtils.cu:36; details: CUDA error 35 CUDA driver version is insufficient for CUDA runtime version 2021-07-28 12:54:41,094 - annb.319d272f67 - ERROR - Faiss assertion 'err__ == cudaSuccess' failed in int faiss::gpu::getNumDevices() at /root/miniconda3/conda-bld/faiss-pkg_1623030479928/work/faiss/gpu/utils/DeviceUtils.cu:36; details: CUDA error 35 CUDA driver version is insufficient for CUDA runtime version 2021-07-28 12:54:41,094 - annb.319d272f67 - ERROR - Child process for container 319d272f67 raised exception 139 with host having NVIDIA-SMI 470.42.01 Driver Version: 470.42.01 CUDA Version: 11.4. Would I have to update the dockerfile to point at more recent version? If it runs in your setup, I'm fine with merging. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#9 (comment)](https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#issuecomment-888220165)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AADL6CJADYMMTXRDE7TAVADTZ7P3VANCNFSM5BBCWJOQ> .

I was missing the nvidia running time for docker, now it works on my setup as well. 

One final detail: it seems that the docker images in `t3/faiss_t3/algos.yaml` have the wrong name. They are missing an underscore before `t3`, right? I had to change this locally. 

I would be happy to merge this into main after this fix. Is there any more work on you want to do in this branch?

gosha1128 commented:
Fantastic.  I’ll make that change.

I’d like to test Mathijs new code for SSNPP before you accept the pull
request.

George

On Fri, Jul 30, 2021 at 7:37 AM Martin Aumüller ***@***.***>
wrote:

> Yeah. I should put the full installation instructions for the FAISS T3
> baseline into the README ( it's just a light mention right now. ) My setup
> is 1) standard CUDA 11.0 drivers, and 2) the CUDA 11.0 docker/container run
> time:
> https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html
> Note that if you want to try without docker (ie, you just have the CUDA
> drivers), I added a "--nodocker" flag to run.py. This was useful during
> debugging. George
> … <#m_-8294087081766308037_>
> On Wed, Jul 28, 2021 at 4:04 AM Martin Aumüller *@*.*> wrote: Running
> this on a local server, I get @.*:/mnt/large_storage/maau/big-ann-benchmarks$
> python3 run.py --dataset deep-10M --algorithm faiss-t3 file
> data/deep1b/query.public.10K.fbin already exists file data/deep1b/deep-10M
> already exists 2021-07-28 12:54:35,235 - annb - INFO - running only
> faiss-t3 2021-07-28 12:54:35,492 - annb - INFO - Order:
> [Definition(algorithm='faiss-t3', constructor='FaissT3',
> module='benchmark.algorithms.faiss_t3',
> docker_tag='billion-scale-benchmark-faiss_t3', arguments=['euclidean',
> {'indexkey': 'IVF65536,SQ8', 'buildthreads': -1, 'by_residual': -1,
> 'add_bs': 100000, 'maxtrain': 0, 'clustering_niter': -1, 'add_splits': 1}],
> query_argument_groups=[['nprobe=1'], ['nprobe=2'], ['nprobe=4'],
> ['nprobe=8'], ['nprobe=16'], ['nprobe=32'], ['nprobe=64'], ['nprobe=128'],
> ['nprobe=256']], disabled=False)] 2021-07-28 12:54:36,152 - annb.319d272f67
> - INFO - Created container 319d272f67: CPU limit 0-55, mem limit
> 535741299456, timeout 43200, command ['--dataset', 'deep-10M',
> '--algorithm', 'faiss-t3', '--module', 'benchmark.algorithms.faiss_t3',
> '--constructor', 'FaissT3', '--runs', '5', '--count', '10', '["euclidean",
> {"indexkey": "IVF65536,SQ8", "buildthreads": -1, "by_residual": -1,
> "add_bs": 100000, "maxtrain": 0, "clustering_niter": -1, "add_splits":
> 1}]', '["nprobe=1"]', '["nprobe=2"]', '["nprobe=4"]', '["nprobe=8"]',
> '["nprobe=16"]', '["nprobe=32"]', '["nprobe=64"]', '["nprobe=128"]',
> '["nprobe=256"]'] 2021-07-28 12:54:40,490 - annb.319d272f67 - INFO - Faiss
> assertion 'err__ == cudaSuccess' failed in int faiss::gpu::getNumDevices()
> at
> /root/miniconda3/conda-bld/faiss-pkg_1623030479928/work/faiss/gpu/utils/DeviceUtils.cu:36;
> details: CUDA error 35 CUDA driver version is insufficient for CUDA runtime
> version 2021-07-28 12:54:41,094 - annb.319d272f67 - ERROR - Faiss assertion
> 'err__ == cudaSuccess' failed in int faiss::gpu::getNumDevices() at
> /root/miniconda3/conda-bld/faiss-pkg_1623030479928/work/faiss/gpu/utils/DeviceUtils.cu:36;
> details: CUDA error 35 CUDA driver version is insufficient for CUDA runtime
> version 2021-07-28 12:54:41,094 - annb.319d272f67 - ERROR - Child process
> for container 319d272f67 raised exception 139 with host having NVIDIA-SMI
> 470.42.01 Driver Version: 470.42.01 CUDA Version: 11.4. Would I have to
> update the dockerfile to point at more recent version? If it runs in your
> setup, I'm fine with merging. — You are receiving this because you were
> mentioned. Reply to this email directly, view it on GitHub <#9 (comment)
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#issuecomment-888220165>>,
> or unsubscribe
> https://github.com/notifications/unsubscribe-auth/AADL6CJADYMMTXRDE7TAVADTZ7P3VANCNFSM5BBCWJOQ
> .
>
> I was missing the nvidia running time for docker, now it works on my setup
> as well.
>
> One final detail: it seems that the docker images in
> t3/faiss_t3/algos.yaml have the wrong name. They are missing an
> underscore before t3, right? I had to change this locally.
>
> I would be happy to merge this into main after this fix. Is there any more
> work on you want to do in this branch?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#issuecomment-889936547>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADL6CMIEKJPE2B6LMWYAR3T2K2KJANCNFSM5BBCWJOQ>
> .
>

maumueller commented:
Could you try to move his code to `faiss_t3.py` if it seems straight forward to do? I can take a stab at it otherwise after the merge.

gosha1128 commented:
Yes that is my plan.

On Fri, Jul 30, 2021 at 10:22 AM Martin Aumüller ***@***.***>
wrote:

> Could you try to move his code to faiss_t3.py if it seems straight
> forward to do? I can take a stab at it otherwise after the merge.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#issuecomment-890040670>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADL6CMXYRYZM3OO5ZGJXSTT2LNV5ANCNFSM5BBCWJOQ>
> .
>

gosha1128 commented:
Alright, the new code is integrated into benchmark/algorithms/faiss_t3:

https://github.com/harsha-simhadri/big-ann-benchmarks/blob/gw/T3/benchmark/algorithms/faiss_t3.py

I tested it on ssnpp-10M.  Now doing 1B.

On Fri, Jul 30, 2021 at 10:37 AM George Williams ***@***.***>
wrote:

>
> Yes that is my plan.
>
> On Fri, Jul 30, 2021 at 10:22 AM Martin Aumüller ***@***.***>
> wrote:
>
>> Could you try to move his code to faiss_t3.py if it seems straight
>> forward to do? I can take a stab at it otherwise after the merge.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#issuecomment-890040670>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AADL6CMXYRYZM3OO5ZGJXSTT2LNV5ANCNFSM5BBCWJOQ>
>> .
>>
>

gosha1128 commented:
> Could you try to move his code to `faiss_t3.py` if it seems straight forward to do? I can take a stab at it otherwise after the merge.

Alright, the new code is integrated into benchmark/algorithms/faiss_t3:

https://github.com/harsha-simhadri/big-ann-benchmarks/blob/gw/T3/benchmark/algorithms/faiss_t3.py

I tested it on ssnpp-10M.  Now doing 1B.

gosha1128 commented:
> > Yeah. I should put the full installation instructions for the FAISS T3 baseline into the README ( it's just a light mention right now. ) My setup is 1) standard CUDA 11.0 drivers, and 2) the CUDA 11.0 docker/container run time: https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html Note that if you want to try without docker (ie, you just have the CUDA drivers), I added a "--nodocker" flag to run.py. This was useful during debugging. George
> 
> I was missing the nvidia-runtime for docker, now it works on my setup as well.
> 
> One final detail: it seems that the docker images in `t3/faiss_t3/algos.yaml` have the wrong name. They are missing an underscore before `t3`, right? I had to change this locally.
> 
> I would be happy to merge this into main after this fix. Is there any more work on you want to do in this branch?

I copied what was done in the main algos.yaml:

docker-tag: billion-scale-benchmark-faissconda

Should I change "billion-scale-benchmark-faisst3" to "billion-scale-benchmark-faiss-t3" ?

maumueller commented:
with the current version it results in
```
$ python3 run.py --definitions t3/faiss_t3/algos.yaml --dataset random-xs --t3
Preparing datasets with 10000 random points and 1000 queries.
Computing groundtruth
2021-07-30 22:19:20,521 - annb - INFO - not all docker images available, only: {'ubuntu', 'ann-benchmarks', 'nvidia/cuda', 'ann-benchmarks-mrpt', 'billion-scale-benchmark', 'billion-scale-benchmark-faiss', 'ann-benchmarks-nmslib', 'ann-benchmarks-kgraph', 'billion-scale-benchmark-faiss_t3', 'ann-benchmarks-mih', 'ann-benchmarks-faissgpu', 'ann-benchmarks-reproducibility-latex', 'ann-benchmarks-faiss', 'ann-benchmarks-rpforest', 'billion-scale-benchmark-faissconda', 'ann-benchmarks-annoy', 'ann-benchmarks-flann', 'ann-benchmarks-pynndescent', 'gputest', 'ann-benchmarks-ngt'}
2021-07-30 22:19:20,522 - annb - INFO - missing docker images: {'billion-scale-benchmark-faisst3'}
Traceback (most recent call last):
  File "run.py", line 6, in <module>
    main()
  File "/mnt/large_storage/maau/big-ann-benchmarks/benchmark/main.py", line 209, in main
    raise Exception('Nothing to run')
Exception: Nothing to run
```

(Note that billion-scale-benchmark-faiss_t3 is available.) So it should indeed be `_t3` at the end.

gosha1128 commented:
You are right.  Looking at my docker images I have both, likely from early
development work.  I'll fix it now.

billion-scale-benchmark-faiss_t3
          latest                   e7a20a3e5c06        4 days ago
 9.57GB
billion-scale-benchmark-faisst3
         latest                   0345996e6c74        5 days ago
 9.56GB

On Fri, Jul 30, 2021 at 1:20 PM Martin Aumüller ***@***.***>
wrote:

> with the current version it results in
>
> $ python3 run.py --definitions t3/faiss_t3/algos.yaml --dataset random-xs --t3
> Preparing datasets with 10000 random points and 1000 queries.
> Computing groundtruth
> 2021-07-30 22:19:20,521 - annb - INFO - not all docker images available, only: {'ubuntu', 'ann-benchmarks', 'nvidia/cuda', 'ann-benchmarks-mrpt', 'billion-scale-benchmark', 'billion-scale-benchmark-faiss', 'ann-benchmarks-nmslib', 'ann-benchmarks-kgraph', 'billion-scale-benchmark-faiss_t3', 'ann-benchmarks-mih', 'ann-benchmarks-faissgpu', 'ann-benchmarks-reproducibility-latex', 'ann-benchmarks-faiss', 'ann-benchmarks-rpforest', 'billion-scale-benchmark-faissconda', 'ann-benchmarks-annoy', 'ann-benchmarks-flann', 'ann-benchmarks-pynndescent', 'gputest', 'ann-benchmarks-ngt'}
> 2021-07-30 22:19:20,522 - annb - INFO - missing docker images: {'billion-scale-benchmark-faisst3'}
> Traceback (most recent call last):
>   File "run.py", line 6, in <module>
>     main()
>   File "/mnt/large_storage/maau/big-ann-benchmarks/benchmark/main.py", line 209, in main
>     raise Exception('Nothing to run')
> Exception: Nothing to run
>
> (Note that billion-scale-benchmark-faiss_t3 is available.) So it should
> indeed be _t3 at the end.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#issuecomment-890133236>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADL6CM3BAT26PZLPXELZNTT2MCSNANCNFSM5BBCWJOQ>
> .
>

gosha1128 commented:
> Running this on a local server, I get
> 
> ```
> maau@preciousss:/mnt/large_storage/maau/big-ann-benchmarks$ python3 run.py --dataset deep-10M --algorithm faiss-t3
> file data/deep1b/query.public.10K.fbin already exists
> file data/deep1b/deep-10M already exists
> 2021-07-28 12:54:35,235 - annb - INFO - running only faiss-t3
> 2021-07-28 12:54:35,492 - annb - INFO - Order: [Definition(algorithm='faiss-t3', constructor='FaissT3', module='benchmark.algorithms.faiss_t3', docker_tag='billion-scale-benchmark-faiss_t3', arguments=['euclidean', {'indexkey': 'IVF65536,SQ8', 'buildthreads': -1, 'by_residual': -1, 'add_bs': 100000, 'maxtrain': 0, 'clustering_niter': -1, 'add_splits': 1}], query_argument_groups=[['nprobe=1'], ['nprobe=2'], ['nprobe=4'], ['nprobe=8'], ['nprobe=16'], ['nprobe=32'], ['nprobe=64'], ['nprobe=128'], ['nprobe=256']], disabled=False)]
> 2021-07-28 12:54:36,152 - annb.319d272f67 - INFO - Created container 319d272f67: CPU limit 0-55, mem limit 535741299456, timeout 43200, command ['--dataset', 'deep-10M', '--algorithm', 'faiss-t3', '--module', 'benchmark.algorithms.faiss_t3', '--constructor', 'FaissT3', '--runs', '5', '--count', '10', '["euclidean", {"indexkey": "IVF65536,SQ8", "buildthreads": -1, "by_residual": -1, "add_bs": 100000, "maxtrain": 0, "clustering_niter": -1, "add_splits": 1}]', '["nprobe=1"]', '["nprobe=2"]', '["nprobe=4"]', '["nprobe=8"]', '["nprobe=16"]', '["nprobe=32"]', '["nprobe=64"]', '["nprobe=128"]', '["nprobe=256"]']
> 2021-07-28 12:54:40,490 - annb.319d272f67 - INFO - Faiss assertion 'err__ == cudaSuccess' failed in int faiss::gpu::getNumDevices() at /root/miniconda3/conda-bld/faiss-pkg_1623030479928/work/faiss/gpu/utils/DeviceUtils.cu:36; details: CUDA error 35 CUDA driver version is insufficient for CUDA runtime version
> 2021-07-28 12:54:41,094 - annb.319d272f67 - ERROR - Faiss assertion 'err__ == cudaSuccess' failed in int faiss::gpu::getNumDevices() at /root/miniconda3/conda-bld/faiss-pkg_1623030479928/work/faiss/gpu/utils/DeviceUtils.cu:36; details: CUDA error 35 CUDA driver version is insufficient for CUDA runtime version
> 
> 2021-07-28 12:54:41,094 - annb.319d272f67 - ERROR - Child process for container 319d272f67 raised exception 139
> ```
> 
> with host having `NVIDIA-SMI 470.42.01 Driver Version: 470.42.01 CUDA Version: 11.4`. Would I have to update the dockerfile to point at more recent version? If it runs in your setup, I'm fine with merging.

I'm running CUDA 11.0 because I have other projects that depend on this version at the moment.  If you would like to change to a more recent version of CUDA, please feel free to test and change the version.

gosha1128 commented:
> You are right. Looking at my docker images I have both, likely from early development work. I'll fix it now. billion-scale-benchmark-faiss_t3 latest e7a20a3e5c06 4 days ago 9.57GB billion-scale-benchmark-faisst3 latest 0345996e6c74 5 days ago 9.56GB
> […](#)
> On Fri, Jul 30, 2021 at 1:20 PM Martin Aumüller ***@***.***> wrote: with the current version it results in $ python3 run.py --definitions t3/faiss_t3/algos.yaml --dataset random-xs --t3 Preparing datasets with 10000 random points and 1000 queries. Computing groundtruth 2021-07-30 22:19:20,521 - annb - INFO - not all docker images available, only: {'ubuntu', 'ann-benchmarks', 'nvidia/cuda', 'ann-benchmarks-mrpt', 'billion-scale-benchmark', 'billion-scale-benchmark-faiss', 'ann-benchmarks-nmslib', 'ann-benchmarks-kgraph', 'billion-scale-benchmark-faiss_t3', 'ann-benchmarks-mih', 'ann-benchmarks-faissgpu', 'ann-benchmarks-reproducibility-latex', 'ann-benchmarks-faiss', 'ann-benchmarks-rpforest', 'billion-scale-benchmark-faissconda', 'ann-benchmarks-annoy', 'ann-benchmarks-flann', 'ann-benchmarks-pynndescent', 'gputest', 'ann-benchmarks-ngt'} 2021-07-30 22:19:20,522 - annb - INFO - missing docker images: {'billion-scale-benchmark-faisst3'} Traceback (most recent call last): File "run.py", line 6, in <module> main() File "/mnt/large_storage/maau/big-ann-benchmarks/benchmark/main.py", line 209, in main raise Exception('Nothing to run') Exception: Nothing to run (Note that billion-scale-benchmark-faiss_t3 is available.) So it should indeed be _t3 at the end. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#9 (comment)](https://github.com/harsha-simhadri/big-ann-benchmarks/pull/9#issuecomment-890133236)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AADL6CM3BAT26PZLPXELZNTT2MCSNANCNFSM5BBCWJOQ> .

OK, I fixed it.
